### ISSUE TITLE
Using Sparse tensors to apply gradients in BackPropogation. 

### ISSUE THREAD

**Issue by raginisharma14** [tensorflow/tensorflow#12772]
Hi All,

I am trying to use sparse tensors while applying gradients and i see below error.
 **Tensor conversion requested dtype int64 for Tensor with dtype float32: 'Tensor("Adam_1/update_conv1_1/weights/sub_2:0", shape=(), dtype=float32)'**

I am assuming this to be a bug as the dtype int64 must have been hardcoded for sparse tensors. 
Is there any workaround for this issue? 

Here is what I am trying to do:
#################### this function multiplies gradients with prune weights so that gradient updation happens on pruned gradients.
```
def apply_prune_on_grads(grads_and_vars, sess, dict_n):
     print("im inside pply_prune_on_grads")
     i=0
     for k, v in dict_n.items():
         count=0
         for grad, var in grads_and_vars:
             if var.name == k:
                 op = (var.name).split("/")
                 if op[1] != 'biases:0':
                     v = tf.cast(tf.constant(v), tf.float32)
                     mask_tensor = tf.multiply(v,grad)
                     idx = tf.where(tf.not_equal(mask_tensor, 0))
                     sparse = tf.IndexedSlices(idx,tf.gather_nd(mask_tensor, idx))
                     grads_and_vars[count] = (sparse, var)
             count = count+1
         i=i+1
     return grads_and_vars

def prune_weights(weights_prune, sess,threshold = 0.01):
         print("im inside prune weights")
         sparse_weights = {}
         for v in weights_prune:
             value = sess.run(v) 
             under_threshold = abs(value) < threshold
             value[under_threshold] = 0
             sess.run(v.assign(value))
             sparse_weights[v.name] = np.logical_not(under_threshold)
         return sparse_weights
```

############################### The order in which i am calling the above functions in sess.run
```
 sparse_weights = prune_weights(variables_to_restore, sess)
 grads_and_vars = train_op.compute_gradients(loss, )
 grads_and_vars =  apply_prune_on_grads(grads_and_vars,sess,  sparse_weights)
 train_step = train_op.apply_gradients(grads_and_vars) 

```

**Comment by raginisharma14** [comment]
I could resolve the issue by typecasting tf.gather_nd(mask_tensor, idx) to tf.to_float(tf.gather_nd(mask_tensor, idx))

**Comment by facaiy** [comment]
Hi, @raginisharma14 . Could you give a more concise (minimal, self-contained) reproduced code?

If I understand correctly, `Tensor conversion requested dtype int64 for Tensor with dtype float32` means that int64 is expected here, not float32. Hence it seems a little surprising for me to use `tf.to_float(xxxx)`.

**Comment by raginisharma14** [comment]
Hi @facaiy , in tf.IndexedSlice , the first argument is values and second argument is indices, But I have passed them in the reverse order, I think that is the reason why i was seeing the error."Tensor conversion requested dtype int64 for Tensor with dtype float32". When I changed it to values, indices. I don't see that error anymore.
but I am still facing the Rank issue while scatter_sub is called in apply_gradients. 
"Shapes must be equal rank, but are 1 and 5 for 'GradientDescent_1/update_conv1_1/weights/ScatterSub' (op: 'ScatterSub') with input shapes: [3,3,3,64], [?,4], [?]"
Is there a way I can calculate gradients on sparse matrices? Is the issue because scatter  sub does not support multidimensional matrix?

**Comment by facaiy** [comment]
Sorry, @raginisharma14 . I am not familiar with SparseTensor, hence the question is beyond my scope. 

By the way,  it would be wonderful if the test code could be more concise.

**Comment by raginisharma14** [comment]
@facaiy : Here is what I am trying to do: 1. I am applying pruning on to the weights of each layer and then multiplying the pruned weights with the computed gradients and then converting the output of(pruned weights * computed gradients) to a sparse matrix(using tf.IndexedSlices). After this, I would apply these gradients. In tensorflow, while applying gradients you can either pass sparse gradients or dense.  But while the scatter_sub is called it is throwing error  Shapes must be equal rank, but are 1 and 5 for 'GradientDescent_1/update_conv1_1/weights/ScatterSub' (op: 'ScatterSub') with input shapes: [3,3,3,64], [?,4], [?]
It's failing while calling scatter_sub in the below line
https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/python/training/gradient_descent.py#L67

**Comment by cuixue** [comment]
I met the same problem with you. and have you solved it?

**Comment by drpngx** [comment]
Yes, I think it's a usage issue. Feel free to continue commenting on the issue, but I am closing this. Please re-open if you think it's a bug in tensorflow.

**Comment by tTianyu** [comment]
Hi! @raginisharma14 @cuixue I met the problem,too . have you fix it ?
